model/histogram/float_histogram.go

```diff
@@ -53,28 +53,57 @@ type FloatHistogram struct {
 
 // Copy returns a deep copy of the Histogram.
 func (h *FloatHistogram) Copy() *FloatHistogram {
-	c := *h
+	c := FloatHistogram{
+		CounterResetHint: h.CounterResetHint,
+		Schema:           h.Schema,
+		ZeroThreshold:    h.ZeroThreshold,
+		ZeroCount:        h.ZeroCount,
+		Count:            h.Count,
+		Sum:              h.Sum,
+	}
 
-	if h.PositiveSpans != nil {
+	if len(h.PositiveSpans) != 0 {
 		c.PositiveSpans = make([]Span, len(h.PositiveSpans))
 		copy(c.PositiveSpans, h.PositiveSpans)
 	}
-	if h.NegativeSpans != nil {
+	if len(h.NegativeSpans) != 0 {
 		c.NegativeSpans = make([]Span, len(h.NegativeSpans))
 		copy(c.NegativeSpans, h.NegativeSpans)
 	}
-	if h.PositiveBuckets != nil {
+	if len(h.PositiveBuckets) != 0 {
 		c.PositiveBuckets = make([]float64, len(h.PositiveBuckets))
 		copy(c.PositiveBuckets, h.PositiveBuckets)
 	}
-	if h.NegativeBuckets != nil {
+	if len(h.NegativeBuckets) != 0 {
 		c.NegativeBuckets = make([]float64, len(h.NegativeBuckets))
 		copy(c.NegativeBuckets, h.NegativeBuckets)
 	}
 
 	return &c
 }
 
+// CopyFrom makes a deep copy of the given FloatHistogram into the receiver object.
+func (h *FloatHistogram) CopyFrom(from FloatHistogram) {
```

Comment:
Hm, depending on what's more consistent/easier we could have only "two" copy methods:

`func (h *FloatHistogram) Copy(buf *FloatHistogram) *FloatHistogram {`
and 
`func (h *FloatHistogram) CopyToSchema(targetSchema int32, buf *FloatHistogram) *FloatHistogram {`

I feel this is what we do in other places in Prometheus e.g. ChunkIterators. 

Additionally I am not sure about extra value copy here in "from FloatHistogram". We already have it on heap so why copying things to (potentially to stack) for this function? Probably no perf difference, just weird? I think if we keep CopyFrom it should be `CopyFrom(from *FloatHistogram`) perhaps 🤔

Reply:
What about "buf" idea and changing existing methods?

Reply:
I feel like the "buf" API could indicate that the passed in histogram could be safely reused after calling Copy, leading to subtle data corruption bugs. Separate `CopyTo` and `CopyToSchema` methods do make sense though, but I don't really have a strong opinion. I see that we already have a `func (ls *Labels) CopyFrom(b Labels)` method though, so not sure what's more consistent with the rest of the codebase.

Reply:
I tried to think through how merging CopyFrom (or more likely CopyTo) with CopyToSchema work. In case the requested schema is equal to the current schema, it's simply an optimization like in AtHistogram. In case the schema is different (lower), then I'd expect to reuse the spans/buckets in the target histogram, which you could do by first copying over the target histogram then executing `reduceResolution` with `inplace==true`. 

So technically it seems possible, but I'm not sure we want to complicate the code and testing as the use cases don't intersect at the moment. You'd want to use CopyTo to fill the samples being input to the engine and then CopyToSchema is used when calculating e.g. rate() in the engine, and there it's unlikely you want reuse.

To summarize: agree with Filip on keeping separate and making simpler `CopyTo`, want to pass pointers everywhere for consistency.   

When looking into this, I was reminded that histogram Add/Sub can change samples in place by altering the schema, which makes reusing those series in other parts of the query execution less precise. Basically we're assuming that we'll never optimize away loading series for each sub query individually :)

Reply:
Sounds good, I will replace `CopyFrom` with `CopyTo` for consistency with other histogram methods.

Reply:
I've replaced CopyFrom with CopyTo, let me know if this looks good now.